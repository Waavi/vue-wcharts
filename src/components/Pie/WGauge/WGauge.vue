<template>
    <g
        v-if="active"
        id="gauge"
    >
        <WPie
            :radius="radius"
            :startAngle="startAngle"
            :endAngle="endAngle"
            :animation="false"
            :color="stylesCmp.background"
            :borderRadius="borderRadius"
        />
        <WPie
            :datakey="datakey"
            :maxValue="maxValue"
            :radius="radius"
            :startAngle="startAngle"
            :endAngle="endAngle"
            :color="color"
            :borderRadius="borderRadius"
            :animation="animation"
            :animationDuration="animationDuration"
        />

        <foreignObject
            v-if="slots"
            :style="contentStyles"
        >
            <slot />
        </foreignObject>
    </g>
</template>

<script>
import VueTypes from 'vue-types'
import themeMixin from '../../../mixins/theme'
import visibleMixin from '../../../mixins/visible'
import WPie from '../WPie/WPie.vue'

export default {
    name: 'WGauge',
    type: 'pie',
    inject: ['Chart'],
    components: {
        WPie,
    },
    mixins: [themeMixin, visibleMixin],
    props: {
        // internal props set by the parent (WPieChart)
        index: VueTypes.number,
        datakey: VueTypes.string.isRequired,
        maxValue: VueTypes.number.def(100),
        trigger: VueTypes.oneOf(['hover', 'click', 'manual']).def('hover'),
        startAngle: VueTypes.number.def(180),
        endAngle: VueTypes.number.def(0),
        radius: VueTypes.oneOfType([
            VueTypes.number,
            VueTypes.arrayOf(VueTypes.number).def([85, 100]),
        ]).def([85, 100]),
        // Styles
        borderRadius: VueTypes.oneOfType([VueTypes.number, VueTypes.string]).def(0),
        color: VueTypes.string,
        styles: VueTypes.object,
        // Animation
        animation: VueTypes.bool.def(true),
        animationDuration: VueTypes.number.def(2.5),
    },
    computed: {
        // Active elem
        active () {
            return this.Chart.activeElements.includes(this.index)
        },
        // Radius
        curRadius () {
            const innerRadius = Array.isArray(this.radius) ? this.radius[0] : 0
            const outerRadius = Array.isArray(this.radius) ? this.radius[1] : 100

            return { innerRadius, outerRadius }
        },
        // Slots
        slots () {
            return Object.keys(this.$slots).length
        },
        // Styles
        contentStyles () {
            const { height } = this.Chart
            const { outerRadius } = this.curRadius
            const size = `${outerRadius * 2}px`
            return {
                width: size,
                height: size,
                transform: `translate(${outerRadius}px, ${height / 2 - outerRadius}px)`,
            }
        },
        stylesCmp () {
            return {
                ...this.themeStyles.styles,
                ...this.styles,
            }
        },
    },
}
</script>
