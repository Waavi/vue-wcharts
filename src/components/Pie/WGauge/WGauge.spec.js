import { shallowMount } from '@vue/test-utils'
import WGauge from './WGauge.vue'

describe('Components/WGauge', () => {
    const propsData = {
        datakey: 'one',
        animation: false,
        index: 0,
    }
    const dataset = [
        {
            name: 'Page A', one: 70, total: 100,
        },
    ]

    const provide = {
        Chart: {
            dataset,
            data: dataset,
            width: 740,
            height: 400,
            cx: '50%',
            cy: '50%',
            startAngle: 0,
            endAngle: 360,
            paddingAngle: 0,
            activeElements: [0],
            colors: ['red', 'red', 'red', 'red', 'red', 'red', 'red'],
            curCx: 300,
            curCy: 200,
            setActive: () => undefined,
            cleanActive: () => undefined,
        },
    }

    const defaultConfig = {
        propsData,
        provide,
    }

    it(`Should be render correctly`, () => {
        const wrapper = shallowMount(WGauge, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render radius correctly`, () => {
        const wrapper = shallowMount(WGauge, defaultConfig)
        expect(wrapper.vm.curRadius).toMatchSnapshot()
    })

    it(`Should be render styles correctly`, () => {
        const wrapper = shallowMount(WGauge, defaultConfig)
        expect(wrapper.vm.contentStyles).toMatchSnapshot()
    })
})
