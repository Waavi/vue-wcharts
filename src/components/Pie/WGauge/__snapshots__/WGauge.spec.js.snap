// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/WGauge Should be render correctly 1`] = `
<g id="gauge">
  <WPie-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" visible="true" index="0" datakey="" trigger="hover" startAngle="180" endAngle="0" radius="85,100" maxValue="-1" styles="[object Object]" color="#eee" borderRadius="0" opacityDisabled="0.5" animationDuration="2.5"></WPie-stub>
  <WPie-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" visible="true" index="0" datakey="one" trigger="hover" startAngle="180" endAngle="0" radius="85,100" maxValue="100" styles="[object Object]" color="" borderRadius="0" opacityDisabled="0.5" animationDuration="2.5"></WPie-stub>
  <!---->
</g>
`;
